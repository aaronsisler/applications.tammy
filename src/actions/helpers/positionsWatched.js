export const addSubscription = (positionId, subscriptionLevel) => ({
    type: 'ADD_SUBSCRIPTION',
    positionId,
    subscriptionLevel
});

export const removeSubscription = (positionId) => ({
    type: 'REMOVE_SUBSCRIPTION',
    positionId
});

export const setPositionsWatched = (positionsWatched) => ({
    type: 'SET_POSITIONS_WATCHED',
    positionsWatched
});

export const setSubscriptionLevel = (positionId, subscriptionLevel) => ({
    type: 'SET_SUBSCRIPTION_LEVEL',
    positionId,
    subscriptionLevel
});
