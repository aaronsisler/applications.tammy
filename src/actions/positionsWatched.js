import database from 'Firebase/firebase';
import {
    addSubscription,
    removeSubscription,
    setPositionsWatched,
    setSubscriptionLevel
} from 'Actions/helpers/positionsWatched';

export const startAddSubscription = (positionId, subscriptionLevel) => (dispatch, getState) => {
    const { uid: userId } = getState().auth;
    return database.ref(`position_watch/${userId}/${positionId}`).update({ subscriptionLevel })
        .then(() => dispatch(addSubscription(positionId, subscriptionLevel)));
}

export const startRemoveSubscription = (positionId) => (dispatch, getState) => {
    const { uid: userId } = getState().auth;
    return database.ref(`position_watch/${userId}/${positionId}`).remove()
        .then(() => dispatch(removeSubscription(positionId)));
}

export const startSetPositionsWatched = () => (dispatch, getState) => {
    const { uid: userId } = getState().auth;

    return database.ref(`position_watch/${userId}`).once('value').then((snapshot) => {
        const positionsWatched = [];
        snapshot.forEach((childSnapshot) => {
            positionsWatched.push({
                positionId: childSnapshot.key,
                ...childSnapshot.val(),
            });
        });

        return dispatch(setPositionsWatched(positionsWatched));
    });
}

export const startSetSubscriptionLevel = (positionId, subscriptionLevel) => (dispatch, getState) => {
    const { uid: userId } = getState().auth;
    return database.ref(`position_watch/${userId}/${positionId}`).update({ subscriptionLevel })
        .then(() => dispatch(setSubscriptionLevel(positionId, subscriptionLevel)));
}



