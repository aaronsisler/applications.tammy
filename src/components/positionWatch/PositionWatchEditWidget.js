import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { startClearPosition } from 'Actions/position';
import {
    startRemovePositionWatch,
    startSetPositionWatchLevel
} from 'Actions/positionsWatched';

export class PositionWatchEditWidget extends React.Component {
    constructor(props) {
        super(props);
    }

    handleSetPositionWatchLevel = (e) => {
        const notificationLevel = e.target.value;
        this.props.startSetPositionWatchLevel(this.props.positionId, notificationLevel)
    }

    handleRemovePositionWatch = () => {
        this.props.startRemovePositionWatch(this.props.positionId);
        this.props.startClearPosition();
    }

    retrievePositionWatchLevel = () => {
        const { positionId, positionsWatched } = this.props;
        const positionData = positionsWatched
            .find((positionWatched) => positionWatched.positionId === positionId);
        return positionData ? positionData.notificationLevel : undefined;
    }

    render() {
        const currentPositionwatchLevel = this.retrievePositionWatchLevel();
        return (
            <div className="position_watch_edit_widget">
                <div className="position_watch_edit_widget__select_wrapper">
                    <div className="position_watch_edit_widget__title">
                        Notification&nbsp;Level:&nbsp;
                    </div>
                    {currentPositionwatchLevel &&
                        <div className="position_watch_edit_widget__select">
                            <select
                                className="select"
                                value={currentPositionwatchLevel}
                                onChange={this.handleSetPositionWatchLevel}
                            >
                                <option value="ALL">ALL</option>
                                <option value="SOME">SOME</option>
                                <option value="REQUIRED">REQUIRED</option>
                            </select>
                        </div>
                    }
                </div>
                <div className="position_watch_edit_widget__removal">
                    <button
                        className="button"
                        onClick={this.handleRemovePositionWatch}
                    >
                        Remove Watch
                    </button>
                </div>
            </div>
        );
    }
}

/* istanbul ignore next */
const mapStateToProps = (state) => ({
    positionId: state.position.positionId,
    positionsWatched: state.positionsWatched,
});

/* istanbul ignore next */
const mapDispatchToProps = (dispatch) => ({
    startClearPosition: () => dispatch(startClearPosition()),
    startRemovePositionWatch: (positionId) => dispatch(startRemovePositionWatch(positionId)),
    startSetPositionWatchLevel: (positionId, notificationLevel) => dispatch(startSetPositionWatchLevel(positionId, notificationLevel)),
});

export default connect(mapStateToProps, mapDispatchToProps)(PositionWatchEditWidget);

PositionWatchEditWidget.propTypes = {
    positionId: PropTypes.string.isRequired,
    positionsWatched: PropTypes.array,
    startClearPosition: PropTypes.func.isRequired,
    startRemovePositionWatch: PropTypes.func.isRequired,
    startSetPositionWatchLevel: PropTypes.func.isRequired,
};
