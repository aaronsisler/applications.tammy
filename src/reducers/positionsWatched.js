const positionsWatchedReducerDefaultState = [];

const positionsWatchedReducer = (state = positionsWatchedReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_POSITIONS_WATCHED':
            return action.positionsWatched;
        case 'ADD_SUBSCRIPTION':
                return [...state, action.positionWatched];
        case 'REMOVE_SUBSCRIPTION':
            return state.filter((positionwatched) => positionwatched.positionId != action.positionId);
        case 'SET_SUBSCRIPTION_LEVEL':
            return state.map((positionWatched) => {
                if (positionWatched.positionId === action.positionId) {
                    return {
                        ...positionWatched,
                        subscriptionLevel: action.subscriptionLevel
                    }
                }
                return positionWatched;
            })
        default:
            return state;
    }
};

export default positionsWatchedReducer;

