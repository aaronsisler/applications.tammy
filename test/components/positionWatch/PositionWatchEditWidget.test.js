import React from 'react';
import { shallow } from 'enzyme';
import { PositionWatchEditWidget } from 'PositionWatch/PositionWatchEditWidget';
import { positionsWatchedStore } from '../../fixtures/positionsWatched';

describe('PositionWatchEditWidget', () => {
    const [positionWatched] = positionsWatchedStore;
    const { positionId, notificationLevel } = positionWatched;
    const startClearPosition = jest.fn();
    const startRemovePositionWatch = jest.fn();
    const startSetPositionWatchLevel = jest.fn();
    let wrapper;

    const buildWrapper = () => {
        wrapper = shallow(
            <PositionWatchEditWidget
                positionId={positionId}
                positionsWatched={[positionWatched]}
                startClearPosition={startClearPosition}
                startRemovePositionWatch={startRemovePositionWatch}
                startSetPositionWatchLevel={startSetPositionWatchLevel}
            />
        );
    };

    beforeEach(() => {
        buildWrapper();
    });

    it('should render PositionWatchEditWidget correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should set the notificationLevel correctly on construction', () => {
        expect(wrapper.find('select').prop('value')).toEqual(notificationLevel);
    });

    describe('when the notification select input is changed', () => {
        it('should call startSetPositionWatchLevel with positionId and new notification level', () => {
            const newNotificationLevel = 'SOME';
            const event = { target: { value: newNotificationLevel } };

            wrapper.find('select').simulate('change', event);

            expect(startSetPositionWatchLevel).toHaveBeenLastCalledWith(positionId, newNotificationLevel);
        });

        it('should set the new notification level correctly', () => {
            const newNotificationLevel = 'SOME';
            expect(wrapper.find('select').prop('value')).toEqual(notificationLevel);

            wrapper.setProps({ positionsWatched: [{ positionId, notificationLevel: newNotificationLevel }] });

            expect(wrapper.find('select').prop('value')).toEqual(newNotificationLevel);
        });
    });

    describe('when the remove notification button is clicked', () => {
        it('should call startRemovePositionWatch with positionId', () => {
            wrapper.find('button').simulate('click');

            expect(startRemovePositionWatch).toHaveBeenLastCalledWith(positionId);
        });

        it('should call startClearPosition', () => {
            wrapper.find('button').simulate('click');

            expect(startClearPosition).toHaveBeenCalled();
        });
    });
});
