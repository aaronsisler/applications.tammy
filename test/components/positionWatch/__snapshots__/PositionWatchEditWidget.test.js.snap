// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositionWatchEditWidget should render PositionWatchEditWidget correctly 1`] = `
<div
  className="position_watch_edit_widget"
>
  <div
    className="position_watch_edit_widget__select_wrapper"
  >
    <div
      className="position_watch_edit_widget__title"
    >
      Notification Level: 
    </div>
    <select
      className="position_watch_edit_widget__select"
      onChange={[Function]}
      value="ALL"
    >
      <option
        value="ALL"
      >
        ALL
      </option>
      <option
        value="SOME"
      >
        SOME
      </option>
      <option
        value="REQUIRED"
      >
        REQUIRED
      </option>
    </select>
  </div>
  <button
    className="position_watch_edit_widget__button"
    onClick={[Function]}
  >
    Remove Watch
  </button>
</div>
`;

exports[`PositionWatchEditWidget should render PositionWatchEditWidget correctly when positions watched does not contain a matching position id 1`] = `
<div
  className="position_watch_edit_widget"
>
  <div
    className="position_watch_edit_widget__select_wrapper"
  >
    <div
      className="position_watch_edit_widget__title"
    >
      Notification Level: 
    </div>
  </div>
  <button
    className="position_watch_edit_widget__button"
    onClick={[Function]}
  >
    Remove Watch
  </button>
</div>
`;
